ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/BeaconStateUtil.java

```diff
@@ -114,6 +119,74 @@ public static BeaconStateWithCache get_genesis_beacon_state(
     return state;
   }
 
+  public static BeaconState initialize_beacon_state_from_eth1(
+      Bytes32 eth1_block_hash, UnsignedLong eth1_timestamp, List<Deposit> deposits) {
+    UnsignedLong genesis_time =
+        eth1_timestamp.minus(
+            eth1_timestamp
+                .mod(UnsignedLong.valueOf(SECONDS_PER_DAY))
+                .plus(UnsignedLong.valueOf(2).times(UnsignedLong.valueOf(SECONDS_PER_DAY))));
+    Eth1Data eth1_data = new Eth1Data(eth1_block_hash, deposits.size());
+    Bytes32 latest_block_header = new BeaconBlockBody().hash_tree_root();
+
+    MerkleTree<Deposit> merkleTree = DepositUtil.generateMerkleTree(deposits);
+    eth1_data.setDeposit_root(merkleTree.getRoot());
+    BeaconState state =
+        BeaconState(genesis_time, UnsignedLong.valueOf(deposits.size()), latest_block_header);
+
+    // Process deposits
+    DepositUtil.applyBranchProofs(merkleTree, deposits);
+    deposits.forEach(
+        deposit -> {
+          BlockProcessor.process_deposit(state, deposit);
+        });
+
+    // Process activations
+    IntStream.range(0, state.getValidator_registry().size())
+        .forEach(
+            index -> {
+              Validator validator = state.getValidator_registry().get(index);
+              UnsignedLong balance = state.getBalances().get(index);
+              UnsignedLong effective_balance =
+                  min(
+                      balance.minus(balance.mod(UnsignedLong.valueOf(EFFECTIVE_BALANCE_INCREMENT))),
+                      UnsignedLong.valueOf(MAX_EFFECTIVE_BALANCE));
+              validator.setEffective_balance(effective_balance);
+
+              if (validator
+                  .getEffective_balance()
+                  .equals(UnsignedLong.valueOf(MAX_EFFECTIVE_BALANCE))) {
+                validator.setActivation_eligibility_epoch(UnsignedLong.valueOf(GENESIS_EPOCH));
+                validator.setActivation_epoch(UnsignedLong.valueOf(GENESIS_EPOCH));
+              }
+            });
+
+    // Populate active_index_roots and compact_committees_roots
+    List<Integer> indices_list =
+        get_active_validator_indices(state, UnsignedLong.valueOf(GENESIS_EPOCH));
+    Bytes32 active_index_root = hash_tree_root(indices_list);
+    Bytes32 committee_root =
+        CrosslinkCommitteeUtil.get_compact_committees_root(state, GENESIS_EPOCH);
+    IntStream.range(0, EPOCHS_PER_HISTORICAL_VECTOR)
+        .forEach(
+            index -> {
+              state.getActive_index_roots().set(index, active_index_root);
+              state.getCompact_committees_roots().set(index, committee_root);
+            });
+    return state;
+  }
+
+  public static boolean is_valid_genesis_state(BeaconState state) {
+    return !(state.getGenesis_time().compareTo(MIN_GENESIS_TIME) >= 0)
```

Comment:
This returns true when state genesis time is less than min genesis time. It should return false.